p2p/pex/pex_reactor.go

```diff
@@ -585,81 +592,65 @@ func (r *PEXReactor) crawlPeersRoutine() {
+	LastCrawled time.Time
 }
 
-// oldestFirst implements sort.Interface for []crawlPeerInfo
-// based on the LastAttempt field.
-type oldestFirst []crawlPeerInfo
-
-func (of oldestFirst) Len() int           { return len(of) }
-func (of oldestFirst) Swap(i, j int)      { of[i], of[j] = of[j], of[i] }
-func (of oldestFirst) Less(i, j int) bool { return of[i].LastAttempt.Before(of[j].LastAttempt) }
-
-// getPeersToCrawl returns addresses of potential peers that we wish to validate.
-// NOTE: The status information is ordered as described above.
-func (r *PEXReactor) getPeersToCrawl() []crawlPeerInfo {
-	var of oldestFirst
+// CrawlPeers will crawl the network looking for new peer addresses.
+// Exposed for testing.
+func (r *PEXReactor) CrawlPeers(addrs []*p2p.NetAddress) {
+	now := time.Now()
 
-	// TODO: be more selective
-	addrs := r.book.ListOfKnownAddresses()
 	for _, addr := range addrs {
-		if len(addr.ID()) == 0 {
-			continue // dont use peers without id
+		peerInfo, ok := r.crawlPeerInfos[addr.ID]
+		if !ok {
+			peerInfo = &crawlPeerInfo{
+				Addr:        addr,
+				LastCrawled: now.Add(-defaultCrawlPeerInterval).Add(-1 * time.Second),
+			}
+			r.crawlPeerInfos[addr.ID] = peerInfo
 		}
 
-		of = append(of, crawlPeerInfo{
-			Addr:        addr.Addr,
-			LastAttempt: addr.LastAttempt,
-			LastSuccess: addr.LastSuccess,
-		})
-	}
-	sort.Sort(of)
-	return of
-}
+		// Do not attempt to connect with peers we recently crawled.
+		if now.Sub(peerInfo.LastCrawled) < defaultCrawlPeerInterval {
+			continue
+		}
 
-// crawlPeers will crawl the network looking for new peer addresses. (once)
-func (r *PEXReactor) crawlPeers() {
-	peerInfos := r.getPeersToCrawl()
+		// Record crawling attempt.
+		peerInfo.LastCrawled = now
+		// XXX: - grows linearly with a number of peers crawled
+		//      - does not get persisted
+		r.crawlPeerInfos[addr.ID] = peerInfo
 
-	now := time.Now()
-	// Use addresses we know of to reach additional peers
-	for _, pi := range peerInfos {
-		// Do not attempt to connect with peers we recently dialed
-		if now.Sub(pi.LastAttempt) < defaultCrawlPeerInterval {
+		err := r.Switch.DialPeerWithAddress(addr, false)
+		if _, ok := err.(p2p.ErrDialingOrExistingAddress); ok {
 			continue
 		}
-		// Otherwise, attempt to connect with the known address
-		err := r.Switch.DialPeerWithAddress(pi.Addr, false)
 		if err != nil {
-			r.book.MarkAttempt(pi.Addr)
+			r.Logger.Error("Dialing failed", "addr", addr, "err", err)
+			markAddrInBookBasedOnErr(addr, r.book, err)
 			continue
 		}
-		// Ask for more addresses
-		peer := r.Switch.Peers().Get(pi.Addr.ID)
+
+		peer := r.Switch.Peers().Get(addr.ID)
 		if peer != nil {
 			r.RequestAddrs(peer)
 		}
 	}
 }
 
-// attemptDisconnects checks if we've been with each peer long enough to disconnect
-func (r *PEXReactor) attemptDisconnects() {
+// AttemptDisconnects checks if we've been with each peer long enough to disconnect
+// Exposed for testing.
+func (r *PEXReactor) AttemptDisconnects() {
```

Comment:
Will this be useful (and used) outside of the `pex` package?

Reply:
No, only in tests.

Reply:
Does it make sense to un-export `AttemptDisconnects` and `CrawlPeers` then?